### PR TITLE
Closes #1547: Hide pocket error page on Try Again

### LINKED ISSUE #1369: firefox:error:pocketconnection is only visible once
Visit Firefox:error:pocketconnection, see try again, tap that and attempt to visit Firefox:error:pocketconnection again. Expected to see an error but get a white screen.

This is causing issues with testing.

### LINKED ISSUE #1547: Pocket megatile error page is never hidden
Once the Pocket megatile error page is shown (if no network connection or if Pocket fails to retrieve any videos), it is never hidden again even on clicking Try Again.

### PR DESCRIPTION
**Pull request by dnarcese**
Also fixes #1369.
### Pull Request checklist

- [x] This PR includes thorough **tests** or an explanation of why it does not
The test turned out to be more complicated and take more investigation then initially thought. Going to keep it separate from this PR in issue #1387 
- [x] This PR includes a **CHANGELOG entry** or does not need one

### CODE REVIEW

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt [hunk 1]
```diff
@@ -225,19 +225,26 @@ class BrowserNavigationOverlay @JvmOverloads constructor(
     fun showMegaTileError() {
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt]: "This PR includes thorough tests or an explanation of why it does not" is checked, but this PR doesn't include any tests or, as far as I can tell, an explanation of why it doesn't.

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt [hunk 2]
```diff
@@ -225,19 +225,26 @@ class BrowserNavigationOverlay @JvmOverloads constructor(
...
             initMegaTile()
             updateOverlayForCurrentState()
-            pocketVideoMegaTileView.requestFocus()
         }
     }
 
+    private fun hideMegaTileError() {
+        pocketVideosContainer.visibility = View.VISIBLE
+        pocketErrorContainer.visibility = View.GONE
+        pocketVideoMegaTileView.requestFocus()
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt]: This isn't an ideal solution, because we're duplicating visibility state between the VM and the fragment.  With this change, state in these two places could go out of sync, leading to difficult to find bugs.  Also this puts very similar code into two different places, complicating the process of making further changes.

A better solution would be to move this into the `pocketViewModel.state` observer in  `BrowserNavigationOverlay#observeForMegaTile`, so that our displayed UI would be modeled directly by our VM state.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 3]
```diff
@@ -131,6 +127,11 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
     }
 
     override fun onLoadingStateChanged(session: Session, loading: Boolean) {
+        if (!loading && session.url == APP_URL_POCKET_ERROR) {
+            browserOverlay?.showMegaTileError()
+            setOverlayVisible(true)
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: This should live in the PocketViewModel, not the view layer.

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt (hunk 1 above)
Comment (dnarcese) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt]: Oops, #1387 would be the test for this.

Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt]: Once this logic is moved into the VM, it should be testable.  It would definitely be preferable to write tests for this now rather than pushing it off to a new issue, and in general we should try to have at least a few tests in each PR.  Not always possible, but a good goal.  LMK if you need a hand testing this, and I can assist.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt (hunk 3 above)
Comment (dnarcese) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: I'm about confused about this- how would this fit into the VM?

Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: @Baron-Severin TODO: overlay refactoring perhaps? not sure about exposing presentation logic in VM

Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: If you look at `BrowserNavigationOverlay#observeForMegaTile`, whenever it receives a `PocketViewModel.State.Error` event it displays an error megatile.  If we handle this logic in the view layer, our state will get out of sync.  (e.g., the VM might be in state `Feed`, but we've changed it by hand in the view layer to display an error).

Since we've moved Pocket over to MVVM, we want our VM to match our display state as closely as possible.  This means that, instead of putting this logic in the view layer, we should find a way to make the VM emit an `Error` object when one needs to be displayed.

Right now, that will mean adding a temporary method to the PocketVM that sends it the current URL.  Once the `SessionRepo` is in master, it can get it from there and we can remove the temporary method.

Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: @psymoon I'm not sure I follow?  Presentation logic is the VM's job.  It's supposed to mirror display state as closely as possible, and in this case that means that `PocketVM.state` should be `Error` whenever an error is displayed.

Comment (psymoon) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: @Baron-Severin yes, I agree 100% on how vs. what to display logic :)

Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Actually, are we sure we need this code at all? As far as I'm aware, the only thing that can put Pocket into an error state is a failed network call, and the repo already handles that case.  Am I missing an edge case?

Comment (dnarcese) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: This part is just for testing- QA wanted a link that would show the error screen for screenshot tests (#1111).

Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: We have a way to push mocked state to the `PocketVideoRepo` directly in Espresso tests now (`FirefoxTestApplication#pushPocketRepoState`).  If @npark-mozilla and @AaronMT are fine with it, I would like to remove the requirement to show a Pocket error on a specific URL, and use that method instead.

Comment (AaronMT) [reply app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Sure. PocketErrorTest.java would need an accompanying update

**Comment by psymoon** [comment]
Digging a little more deeper into this issue, it looks like `Pocket` should utilize the usage of [ConnectivityManager](https://developer.android.com/reference/android/net/ConnectivityManager). Only try pocket request if there's internet connection available. `Try Again` button also seems redundant since we are constantly trying to check for feed anyway (unless I'm mistaken). Calling `requestFocus()` within LiveData observer seems highly unstable and prone to race condition (especially with pocket request retries). This could be addressed further with `FocusRepo` implementation but for now I think we can force all focus to default if focus is lost. We could also look into different polling algorithms. What do you think, @Baron-Severin @dnarcese?

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt [hunk 4]
```diff
@@ -225,6 +225,7 @@ class BrowserNavigationOverlay @JvmOverloads constructor(
     fun showMegaTileError() {
         pocketVideosContainer.visibility = View.GONE
         pocketErrorContainer.visibility = View.VISIBLE
+
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt]: (Copying @psymoon's comment into a review so we can have a discussion thread)

> Digging a little more deeper into this issue, it looks like Pocket should utilize the usage of ConnectivityManager. Only try pocket request if there's internet connection available. Try Again button also seems redundant since we are constantly trying to check for feed anyway (unless I'm mistaken). Calling requestFocus() within LiveData observer seems highly unstable and prone to race condition (especially with pocket request retries). This could be addressed further with FocusRepo implementation but for now I think we can force all focus to default if focus is lost. We could also look into different polling algorithms. What do you think, @Baron-Severin @dnarcese?

Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/BrowserNavigationOverlay.kt]: > Pocket should utilize the usage of ConnectivityManager. Only try pocket request if there's internet connection available.

I don't have particularly strong feelings on this.  If the user has no internet connection, the requests instantly fail and otherwise have no impact on the user experience.  I would vote against the additional code complexity, but again, not a strong opinion.

> Try Again button also seems redundant since we are constantly trying to check for feed anyway (unless I'm mistaken).

It's definitely a narrow use case.  The backoff will eventually wait for a decent length of time between retries (the wait doubles between each attempt), it could be possible for a user to mess with their connection and then want to retry immediately.  The decision has UX ramifications as well, @aminalhazwani do you have any thoughts here?

> Calling requestFocus() within LiveData observer seems highly unstable and prone to race condition (especially with pocket request retries).

TBF, the requestFocus is within a click listener, which is a lot less concerning.

> We could also look into different polling algorithms.

For sure.  Also, the retry wait starts at 1 second, which is probably a little bit too eager.  There's an idea in A-C about breaking out a generic HTTP API client that handles this kind of stuff, so it would be good to refine our approach before writing that.  https://github.com/mozilla-mobile/android-components/issues/1508
